### ISSUE TITLE
[Cinnamon] Keyboard settings window does not tile left/right

### ISSUE THREAD

**Issue by spamegg1** [linuxmint/mint21-beta#77]
**Describe the bug**
Keyboard settings window does not tile left/right but it can tile up/down.

**To Reproduce**
1. Go to 'Menu/Keyboard'
2. Press Super+Left arrow or Super+Right arrow
3. See error. Nothing happens. (A tiny flicker can be seen on the edges.)

**Expected behavior**
It should tile to the left/right sides of the screen with Super+Left/Right.

**Screenshots**
Here is a video of both correct and wrong behavior:
[cinnamon-2022-07-16T234256+0300.webm](https://user-images.githubusercontent.com/4255997/179371517-707ccd1d-96d3-4511-a122-450af59efee2.webm)

**Hardware and mintUpdate**
* `inxi -Fxxxrz` 
```
 ➜ inxi -Fxxxrz
System:
  Kernel: 5.15.0-41-generic x86_64 bits: 64 compiler: gcc v: 11.2.0
    Desktop: Cinnamon 5.4.4 tk: GTK 3.24.33 vt: 7 dm: LightDM 1.30.0
    Distro: Linux Mint 21 Vanessa base: Ubuntu 22.04 jammy
Machine:
  Type: Desktop Mobo: Gigabyte model: B450M S2H serial: <superuser required>
    BIOS: American Megatrends LLC. v: F62d date: 10/13/2021
CPU:
  Info: quad core model: AMD Ryzen 5 3400G with Radeon Vega Graphics bits: 64
    type: MT MCP smt: enabled arch: Zen/Zen+ note: check rev: 1 cache:
    L1: 384 KiB L2: 2 MiB L3: 4 MiB
  Speed (MHz): avg: 1355 high: 2007 min/max: 1400/3700 boost: enabled
    cores: 1: 1255 2: 1237 3: 1323 4: 1254 5: 1256 6: 1256 7: 2007 8: 1254
    bogomips: 59088
  Flags: avx avx2 ht lm nx pae sse sse2 sse3 sse4_1 sse4_2 sse4a ssse3 svm
Graphics:
  Device-1: AMD Picasso/Raven 2 [Radeon Vega Series / Radeon Mobile Series]
    vendor: Gigabyte driver: amdgpu v: kernel pcie: speed: 8 GT/s lanes: 16
    ports: active: DVI-D-1 empty: DP-1,HDMI-A-1 bus-ID: 0a:00.0
    chip-ID: 1002:15d8 class-ID: 0300
  Display: x11 server: X.Org v: 1.21.1.3 driver: gpu: amdgpu
    note:  X driver n/a display-ID: :0 screens: 1
  Screen-1: 0 s-res: 1920x1080 s-dpi: 96 s-size: 508x285mm (20.0x11.2")
    s-diag: 582mm (22.9")
  Monitor-1: DVI-D-0 mapped: DVI-D-1 model: AOC 2476W serial: <filter>
    res: 1920x1080 hz: 60 dpi: 94 size: 521x293mm (20.5x11.5")
    diag: 598mm (23.5") modes: max: 1920x1080 min: 720x400
  OpenGL: renderer: AMD Radeon Vega 11 Graphics (raven LLVM 13.0.1 DRM 3.42
  5.15.0-41-generic)
    v: 4.6 Mesa 22.0.1 direct render: Yes
Audio:
  Device-1: AMD Raven/Raven2/Fenghuang HDMI/DP Audio driver: snd_hda_intel
    v: kernel pcie: speed: 8 GT/s lanes: 16 bus-ID: 0a:00.1 chip-ID: 1002:15de
    class-ID: 0403
  Device-2: AMD Family 17h HD Audio vendor: Gigabyte driver: snd_hda_intel
    v: kernel pcie: speed: 8 GT/s lanes: 16 bus-ID: 0a:00.6 chip-ID: 1022:15e3
    class-ID: 0403
  Sound Server-1: ALSA v: k5.15.0-41-generic running: yes
  Sound Server-2: PulseAudio v: 15.99.1 running: yes
  Sound Server-3: PipeWire v: 0.3.48 running: yes
Network:
  Device-1: Realtek RTL8111/8168/8411 PCI Express Gigabit Ethernet
    vendor: Gigabyte driver: r8169 v: kernel pcie: speed: 2.5 GT/s lanes: 1
    port: f000 bus-ID: 08:00.0 chip-ID: 10ec:8168 class-ID: 0200
  IF: enp8s0 state: up speed: 100 Mbps duplex: full mac: <filter>
Drives:
  Local Storage: total: 1.61 TiB used: 681.01 GiB (41.4%)
  ID-1: /dev/nvme0n1 vendor: Kingston model: SA2000M8500G size: 465.76 GiB
    speed: 31.6 Gb/s lanes: 4 type: SSD serial: <filter> rev: S5Z42105
    temp: 32.9 C scheme: MBR
  ID-2: /dev/sda vendor: Samsung model: SSD 850 EVO 500GB size: 465.76 GiB
    speed: 6.0 Gb/s type: SSD serial: <filter> rev: 2B6Q scheme: MBR
  ID-3: /dev/sdb vendor: Crucial model: M4-CT256M4SSD2 size: 238.47 GiB
    speed: 6.0 Gb/s type: SSD serial: <filter> rev: 070H scheme: MBR
  ID-4: /dev/sdc vendor: Crucial model: CT512MX100SSD1 size: 476.94 GiB
    speed: 6.0 Gb/s type: SSD serial: <filter> rev: MU02 scheme: MBR
Partition:
  ID-1: / size: 457.38 GiB used: 105.42 GiB (23.0%) fs: ext4
    dev: /dev/nvme0n1p1
Swap:
  Alert: No swap data was found.
Sensors:
  System Temperatures: cpu: 16.8 C mobo: N/A gpu: amdgpu temp: 36.0 C
  Fan Speeds (RPM): N/A
Repos:
  Packages: apt: 2884
  No active apt repos in: /etc/apt/sources.list
  Active apt repos in: /etc/apt/sources.list.d/additional-repositories.list
    1: deb https://typora.io/linux ./
  Active apt repos in: /etc/apt/sources.list.d/official-package-repositories.list
    1: deb http://packages.linuxmint.com vanessa main upstream import backport
    2: deb http://archive.ubuntu.com/ubuntu jammy main restricted universe multiverse
    3: deb http://archive.ubuntu.com/ubuntu jammy-updates main restricted universe multiverse
    4: deb http://archive.ubuntu.com/ubuntu jammy-backports main restricted universe multiverse
    5: deb http://security.ubuntu.com/ubuntu/ jammy-security main restricted universe multiverse
  Active apt repos in: /etc/apt/sources.list.d/opera-stable.list
    1: deb https://deb.opera.com/opera-stable/ stable non-free
  Active apt repos in: /etc/apt/sources.list.d/sbt.list
    1: deb https://repo.scala-sbt.org/scalasbt/debian all main
  Active apt repos in: /etc/apt/sources.list.d/steam.list
    1: deb [arch=amd64,i386] https://repo.steampowered.com/steam/ stable steam
  Active apt repos in: /etc/apt/sources.list.d/vscode.list
    1: deb [arch=amd64] http://packages.microsoft.com/repos/code stable main
Info:
  Processes: 295 Uptime: 5h 41m wakeups: 0 Memory: 13.58 GiB
  used: 3.13 GiB (23.1%) Init: systemd v: 249 runlevel: 5 Compilers:
  gcc: 11.2.0 alt: 11 clang: 14.0.0-1ubuntu1 Shell: Bash v: 5.1.16
  running-in: gnome-terminal inxi: 3.3.13

```
* Is your system all up-to-date ? **YES.**

**Additional context**

**Comment by JosephMcc** [comment]
For a window to be able tile, it has to be able to be resized to fit within that tile zone. A lot of things can effect that. Monitor size and resolution, language being used (can change the min width of strings), and even the theme being used.

Can you tile to the sides by dragging the window? Is it only a problem with the hot keys? If you can't tile with either method then it's simply a limit of the window vs. the screen resolution.

**Comment by SebastJava** [comment]
> Can you tile to the sides by dragging the window?

No.

> If you can't tile with either method then it's simply a limit of the window vs. the screen resolution.

That's what i was thinking. On VM, i was on 1366x768, and it didn't work. No mouse dragging or keyboard shortcut. Then i switched to FHD, 1920x1080, and it all works: east, west, up, down, and 1/4 corners...

But... there is only a very small residual design issue... EDIT: different, inner issue now moved into #83

**Comment by SebastJava** [comment]
@spamegg1 Oh, sorry, i am not totally sure about this... Looking at your specs, you do have a 1920x1080 screen? But it feels smaller, but nice, when looking at your video. So... I guess you have some scaling turned on... I mean, it's like having a small display, but with nice, smooth curves...

**Comment by JosephMcc** [comment]
@SebastJava The problem you are showing is something different. The window thinks it's tiled left in your case. You'll notice the window corners are also squared. But you are also testing in a VM so that throws in another variable.

**Comment by SebastJava** [comment]
@JosephMcc Yes, i know. I was just saying this is probably not an issue, as you already explained, but while doing this test i found this other small issue...

Should i copy this comment, paste it into a new issue, and delete it from here?

**Comment by JosephMcc** [comment]
@SebastJava It should probably be it's own issue. It's certainly a different problem. I'd be curious if anyone can reproduce it outside of a VM.

**Comment by SebastJava** [comment]
> I'd be curious if anyone can reproduce it outside of a VM.

It is right here in the OP's video. Starting from 0:04 you can see it alternating from square to round corners several times. But it is much more subtle than what you get with Mint-Y, of course.

**Comment by spamegg1** [comment]
> Can you tile to the sides by dragging the window?

Nope. Neither with the hot keys, nor with dragging with the mouse. The issue is consistent with both methods. Can tile up/down but not left/right. Here is a video:
[cinnamon-2022-07-17T080951+0300.webm](https://user-images.githubusercontent.com/4255997/179384813-af9eff33-3486-42b6-875e-5e1a1adb307e.webm)

> For a window to be able tile, it has to be able to be resized to fit within that tile zone. A lot of things can effect that. Monitor size and resolution, language being used (can change the min width of strings), and even the theme being used.

OK... but this is the only exception that I found. Pretty much everything else on my system can tile in any direction (including other Setting windows of all kinds). In fact they tile fine even when part of the window is outside the view. Here is a video:
[cinnamon-2022-07-17T081331+0300.webm](https://user-images.githubusercontent.com/4255997/179384893-3241dc37-0f85-4710-a3a3-4723d71dc5ea.webm)

**Comment by spamegg1** [comment]
> @spamegg1 Oh, sorry, i am not totally sure about this... Looking at your specs, you do have a 1920x1080 screen? But it feels smaller, but nice, when looking at your video. So... I guess you have some scaling turned on... I mean, it's like having a small display, but with nice, smooth curves...

I don't have user interface scaling in the Display tab, but I have 1.5x font scaling in the Font Selection tab.

Wow! The issue is gone if I reset Text scaling factor to 1.0, 1.1 or 1.2! 1.3 or anything higher prevents tiling left/right. What the hell... @JosephMcc I guess you were right.

I've always used Text scaling factor with 1.4x or 1.5x on Mint 20.3 and this issue did not exist. I guess this whole new Mutter thing has some new parameters for when tiling happens? Also it's only happening with the Keyboard settings tab. What's so special about the size of the Keyboard settings tab?

**Comment by JosephMcc** [comment]
Well upping the text scaling can make the text strings require a greater horizontal width which prevents the tiling. It just means there is at least one thing in the keyboard settings window that prevents the window from tiling. It will happen with other windows as well, you just haven't come across them. This really isn't a bug. It's just how tiling works. If a windows minimum width is more than half the screen, it simply won't tile.

**Comment by spamegg1** [comment]
I still think there is a bug.

Now I reset Text Scaling factor to 1.0x, and found that the Calculator app will tile left/right, but won't tile up/down. Here is a video:
[cinnamon-2022-07-17T093859+0300.webm](https://user-images.githubusercontent.com/4255997/179387140-1f8bfd82-cb7b-43cf-a841-5e1a49f8cc10.webm)
(It also does not tile up/down by dragging with the mouse.)

I can't really describe it, the bug is very inconsistent. Depending on the order of the hotkey inputs, sometimes it won't tile left/right either, and it will tile up/down. I'll try to come up with an exact sequence of inputs and try to record it.

**Comment by spamegg1** [comment]
In this video Text scaling factor is at 1.0x:
[cinnamon-2022-07-17T095156+0300.webm](https://user-images.githubusercontent.com/4255997/179387529-7980ad28-e231-416f-833c-8e9c097390ce.webm)

and the sequence of inputs are as follows:

1. Super+Right (tiling sound effect plays, it correctly tiles to right)

2. Super+Up (tiling sound effect plays, it incorrectly goes back to center instead of top-right)

3. Super+Left (tiling sound effect plays, nothing happens)

4. Super+Left (tiling sound effect plays, nothing happens)

5. Super+Left (tiling sound effect does not play, nothing happens)

6. Super+Down (tiling sound effect plays, it incorrectly tiles to left side instead of down)

7. Super+Down (tiling sound effect plays, it incorrectly goes back to center instead of down)

8. Super+Right (tiling sound effect plays, nothing happens)

9. Super+Right (tiling sound effect plays, nothing happens)

10. Super+Right (tiling sound effect does not play, nothing happens)

So I think this is what's happening:

1. it thinks it tiled to the right half of the screen, and it displays this correctly.
2. it thinks it tiled to the top-right quadrant of the screen, but it does not display this correctly.
3. it thinks it moved to top half of the screen, but does not display it correctly.
4. it thinks it moved to top-left quadrant of the screen, but does not display it correctly.
5. it thinks that there is no more possibility of tiling to the left, because it's already at the top-left quadrant.
6. it thinks that it moved to the left half of the screen, and it displays this correctly.
7. it thinks it moved to the bottom-left quadrant of the screen, but does not display this correctly.
8. it thinks it moved to the bottom half of the screen, but does not display this correctly.
9. it thinks it moved to the bottom-right quadrant of the screen, but does not display this correctly.
10. it thinks that there is no more possibility of tiling to the right, because it's already at the bottom-right quadrant.

**Comment by SebastJava** [comment]
**For the specific case** of `cinnamon-settings keyboard` I did this simple test:
1. VirtualBox > View menu > Full-screen mode (Host+F) = 1920x1080px here on my screen
2. Font selection > text scaling factor: 1.5
3. Open Keyboard window and try to hand resize the window to 1/2 L or less (<= 960px)
4. Result: the window can't be resized to half the screen width because it has a higher min. setting somewhere...

Now, at first sight, i agree to say this still feels weird, because there isn't much text getting re-scaled there. But just click on the other tabs in this same window and then you'll see this makes sense.

**Comment by JosephMcc** [comment]
@spamegg1 I'm not sure how else to describe this. The calculator won't tile to the top because it requires a minimum height of more than half the height of your screen. It will tile to the side because the window allows itself to fit within the required size. That isn't a bug.

**Comment by SebastJava** [comment]
> I can't really describe it, the bug is very inconsistent.

I half-confirm this. I can find some part of explanation by measuring your calculator in advanced mode and see it is just a bit too tall to get tiled 1/2 top or 1/2 bottom (Super+Up or Super+Down). But that does not explain everything.

**I did some similar tests:**
1. VirtualBox > View menu > Full-screen mode (Host+F) = 1920x1080px here on my screen
2. No font or display scaling
3. Open calculator and set it to Advanced mode
4. Tile left: OK
5. Tile right: OK
6. Tile up: not allowed (normal)
7. Tile left: cancelled for some unknown reason... 2nd attempt... NO... 3rd attempt... NO... Tile right: NO, NO and NO.

To sum up, once you tried to tile up or down, and naturally were not allowed, then nothing else works until you just close and re-open...

P.S.: Of course, if you set your calculator to Normal mode, and make sure it is as small as possible, then it all works fine again. Except... I never got it to get truly centered, even after having clicked that "Center new windows" in `cinnamon-settings windows`. But that's another issue: #75

**Comment by JosephMcc** [comment]
@SebastJava That issue may be linked to your other with tiling with the keyboard. There is a fix in muffin now for that one and I can't reproduce the problem you are describing here now.

Not all apps will follow that Center new windows setting. Some save and restore there placement on their own and override that. The calculator is one of those. Open it, move it somewhere and close it. When you reopen it, it will be in the position you closed it. We have no control over that.

**Comment by spamegg1** [comment]
@JosephMcc If that's the case, then why does it still play the tiling sound effect as if it successfully tiled? 

Why does it do the weird flickering as if it's "trying to tile but failing"? Is there an actual animation programmed for that?

Why does it stop playing the tiling sound when, it would have been in a position unable to tile that way, if it had been tiling correctly up to that point? (Steps 5 and 10 above) 

It's like, even though the window isn't moving, its "tiling history" or whatever is still being kept track of, as if it had tiled. After it fails to tile, in order to tile it some other way, I have to visualize its location and remember where it "would be." (Again please check the sequence of inputs I wrote above.)

I understand what you're saying, but from a user's perspective it absolutely looks, feels and SOUNDS like a bug. 

I would be OK with it if it never moved, played a sound, or flickered, and did not interfere with the further tiling attempts. Then it would communicate to me correctly that "hey I can't tile the way you asked me to, because reasons." 

Anyway, I'll close this since you made it clear it's not a bug.

**Comment by JosephMcc** [comment]
@spamegg1 Your original issue that the keyboard settings wouldn't tile is not a bug and I think I explained why pretty well. The other issues I think are linked to another bug: https://github.com/linuxmint/mint21-beta/issues/83

Tiling with the keyboard keys was broken in some cases so the window thinks it's being tiled. That would lead to the other things like the sound being played and the other problems you are seeing. That will hopefully be fixed now in the next release.

**Comment by spamegg1** [comment]
Ah OK my bad, I guess I misunderstood. It's a complicated situation.

On Mint 20.3 I would always tile my Calculator app to the bottom right quadrant without any problems even with high Text scaling factors (so often that it became muscle memory for me.)

**Comment by JosephMcc** [comment]
I get it. The Calculator app isn't provided by us and was redesigned upstream. Those kinds of changes always take a bit of getting used to.

**Comment by SebastJava** [comment]
> On Mint 20.3 I would always tile my Calculator app to the bottom right quadrant without any problems even with high Text scaling factors...

When this Calculator is set to **Basic mode,** then i can't find any bug. I can do this up, left, right, bottom, 1/4... whatever i want.

On 1920x1080px display with text scaling set to 1.5... :)

**Comment by spamegg1** [comment]
@JosephMcc Right... I guess I shouldn't assume anything or jump to conclusions. It's extremely difficult to tell bugs and non-bugs apart with upstream redesigns like this, when taking old habits into account. Both situations felt like "the same bug" to me. 

I think many other users will be surprised once Mint 21 releases, and might think it's a bug when their habits don't work as before. I guess it's to be expected.

@SebastJava Can confirm the Basic mode, works even with Text scaling. 

I managed to get it tiled up in Advanced Mode by accident, not sure how I did it:
![Screenshot from 2022-07-17 18-40-53](https://user-images.githubusercontent.com/4255997/179405908-5e547001-63bc-42b7-ae87-9a0ad01602ce.png)

As Joseph said, they must have fundamentally changed some things about the app upstream.
